src/hiprtc.cpp

```diff
@@ -0,0 +1,601 @@
+        {
+            std::experimental::filesystem::remove_all(path_);
+        }
+
+        // MANIPULATORS
+        Unique_temporary_path& operator=(
+            const Unique_temporary_path&) = default;
+        Unique_temporary_path& operator=(Unique_temporary_path&&) = default;
+
+        // ACCESSORS
+        const std::experimental::filesystem::path& path() const noexcept
+        {
+            return path_;
+        }
+    };
+} // Unnamed namespace.
+
+namespace hip_impl
+{
+    inline
+    std::string demangle(const char* x)
+    {
+        if (!x) return {};
+
+        int s{};
+        std::unique_ptr<char, decltype(std::free)*> tmp{
+            abi::__cxa_demangle(x, nullptr, nullptr, &s), std::free};
+
+        if (s != 0) return {};
+
+        return tmp.get();
+    }
+} // Namespace hip_impl.
+
+namespace
+{
+    const std::string& defaultTarget()
+    {
+        using namespace std;
+
+        static string r{"gfx900"};
+        static once_flag f{};
+
+        call_once(f, []() {
+            static hsa_agent_t a{};
+            hsa_iterate_agents([](hsa_agent_t x, void*) {
+                hsa_device_type_t t{};
+                hsa_agent_get_info(x, HSA_AGENT_INFO_DEVICE, &t);
+
+                if (t != HSA_DEVICE_TYPE_GPU) return HSA_STATUS_SUCCESS;
+
+                a = x;
+
+                return HSA_STATUS_INFO_BREAK;
+            }, nullptr);
+
+            if (!a.handle) return;
+
+            hsa_agent_iterate_isas(a, [](hsa_isa_t x, void*){
+                uint32_t n{};
+                hsa_isa_get_info_alt(x, HSA_ISA_INFO_NAME_LENGTH, &n);
+
+                if (n == 0) return HSA_STATUS_SUCCESS;
+
+                r.resize(n);
+                hsa_isa_get_info_alt(x, HSA_ISA_INFO_NAME, &r[0]);
+
+                r.erase(0, r.find("gfx"));
+
+                return HSA_STATUS_INFO_BREAK;
+            }, nullptr);
+        });
+
+        return r;
+    }
+
+    inline
+    void handleTarget(std::vector<std::string>& args)
+    {
+        using namespace std;
+
+        bool hasTarget{false};
+        for (auto&& x : args) {
+            const auto dx{x.find("--gpu-architecture")};
+            const auto dy{(dx == string::npos) ? x.find("-arch")
+                                               : string::npos};
+
+            if (dx == dy) continue;
+
+            x.replace(0, x.find('=', min(dx, dy)), "--amdgpu-target");
+            hasTarget = true;
+
+            break;
+        }
+        if (!hasTarget) args.push_back("--amdgpu-target=" + defaultTarget());
+    }
+} // Unnamed namespace.
+
+hiprtcResult hiprtcCompileProgram(hiprtcProgram p, int n, const char** o)
+{
+    using namespace std;
+
+    if (n && !o) return HIPRTC_ERROR_INVALID_INPUT;
+    if (!isValidProgram(p)) return HIPRTC_ERROR_INVALID_PROGRAM;
+    if (p->compiled) return HIPRTC_ERROR_COMPILATION;
+
+    static const string hipcc{
+        getenv("HIP_PATH") ? (getenv("HIP_PATH") + string{"/bin/hipcc"})
```

Comment:
We should probably look for hipcc relative to libhiprtc.so. This is because HIP unit tests don't set env var HIP_PATH and we usually don't install to /opt/rocm when building unit tests. So these tests will fail to find the right hipcc.
Maybe we can do something like (sanity checks/error handling not included):
```cpp
#include <dlfcn.h>
Dl_info dl_info;
dladdr((void*)hiprtcCompileProgram, &dl_info);
string path(dl_info.dli_fname); // path = /path/to/hip/lib/libhiprtc.so
path = path.substr(0, path.find_last_of("/")).substr(0, path.find_last_of("/")); // path = /path/to/hip
static const string hipcc{ path + string{"/bin/hipcc"}};
```

Reply:
I am against this. HIPRTC should just work with any `hipcc`, there is no such thing as the right `hipcc`. Moreover, the user could decide to interpose a flavour of `libhiprtc.so` that lives wherever he / she chooses, and then bad things happen. The test itself should be changed to set HIP_PATH, this alternative is more like the tail wagging the dog.

Reply:
Technically you are right. However here is the issue. HIP runtime has a close relationship with the HCC version. To ensure we are using the right HCC, hipcc has a strict check to confirm that the HCC version that was used to build HIP runtime and the HCC version that is currently being used is the same. So hipcc in another install location could be expecting a different HCC version than the hipcc in the path relative to the current libhiprtc.so. When that happens, hipcc will throw a nice little warning about mismatching HCC versions and abort compilation. So the test will fail.
As regards setting HIP_PATH, https://cmake.org/cmake/help/latest/command/add_test.html does not allow that. So I need to set the path manually *before* I build HIP itself. I don't think that is right either.

Reply:
Yeah that's nice but irrelevant, since the HIPRTC thing is supposed to work with any hipcc, and the latter either works with the environmental hcc or doesn't. It doesn't expect anything in that regard. 

Here's the basic idea, since for some reason it was not crisp enough initially: the .so could be anywhere in the path, with no obvious relationship to the placement of `hipcc`, and that's a valid use case. You cannot rely on its placement to mean anything, and you should not enshrine any implied expectations by rapport with its meaning.

Finally, just for the sake of argument, it is trivial to set the environment for a test, via `set_tests_properties`  + `ENVIRONMENT`, see https://cmake.org/cmake/help/latest/command/set_tests_properties.html and https://cmake.org/cmake/help/latest/prop_test/ENVIRONMENT.html.